 <h4> Tools_for_Data_Science-Assignment - Week 7 Introduction to Pandas (Basic) </h4>
 - handle Read and Write File  <br>
 - handle Data Type and Selection

 ลิงค์ดูคำสั่งเพิ่มเติม
 - https://pandas.pydata.org/docs/user_guide/index.html#user-guide
 - คำสั่งใน pandas ที่ใช้ในสัปดาห์นี้ (แนะนำ) https://colab.research.google.com/drive/1GIhKTKxk1Foa050Ri4QCqQpV9_C-89Ut?usp=sharing
  

In [ ]:
# import pandas json (numpy)

<b> ข้อที่ 1 </b>
1. ให้ดึงข้อมูล CSV จากไฟล์ที่มีชื่อว่า hw-raman_ratings.csv โดยที่มีชื่อคอลัมภ์ดังต่อไปนี้ Review #, Brand, Variety, Style, Country, Stars, Top Ten ตามลำดับ ลิงค์ข้อมูล
<code> https://raw.githubusercontent.com/ABChamp/python_programming_buu_class/master/data/hw-remen-ratings.csv </code>
2. ให้เป็นแปลง string ที่เป็นในรูปแบบ json เป็น Dataframe
3. นำมาข้อมูล Dataframe ที่มาจาก CSV จากไฟล์และ json มาต่อท้ายกัน 
4. เซฟไฟล์เป็น CSV โดยไม่มีมี index ติดมาด้วย (แสดงผลผ่าน stdout)

<b> ผลลัพธ์ </b>
ขนาดของ Dataframe ต้องแถวเท่ากับ 2580 แถวและ 7 คอลัมภ์

<img src="https://drive.google.com/uc?export=view&id=1hBi-d7HuPdZk-2W49rKSaPpvr8zz0Ky3" width="500">

In [ ]:
data = open("https://raw.githubusercontent.com/ABChamp/python_programming_buu_class/master/data/hw-remen-ratings.csv")
# press check the seperator of this file : https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

In [ ]:
jsonString = """[
  {"Review #":10,"Brand":"Smack","Variety":"Vegetable Beef","Style":"Pack","Country":"USA","Stars":"1.5","Top Ten":null},
  {"Review #":9,"Brand":"Sutah","Variety":"Cup Noodle","Style":"Cup","Country":"South Korea","Stars":"2","Top Ten":null},
  {"Review #":8,"Brand":"Tung-I","Variety":"Chinese Beef Instant Rice Noodle","Style":"Pack","Country":"Taiwan","Stars":"3","Top Ten":null},
  {"Review #":7,"Brand":"Ve Wong","Variety":"Mushroom Pork","Style":"Pack","Country":"Vietnam","Stars":"1","Top Ten":null},
  {"Review #":6,"Brand":"Vifon","Variety":"Nam Vang","Style":"Pack","Country":"Vietnam","Stars":"2.5","Top Ten":null},
  {"Review #":5,"Brand":"Vifon","Variety":"Hu Tiu Nam Vang Phnom Penh style Asian Style Instant\u00a0Rice\u00a0Noodles","Style":"Bowl","Country":"Vietnam","Stars":"3.5","Top Ten":null},
  {"Review #":4,"Brand":"Wai Wai","Variety":"Oriental Style Instant Noodles","Style":"Pack","Country":"Thailand","Stars":"1","Top Ten":null},
  {"Review #":3,"Brand":"Wai Wai","Variety":"Tom Yum Shrimp","Style":"Pack","Country":"Thailand","Stars":"2","Top Ten":null},
  {"Review #":2,"Brand":"Wai Wai","Variety":"Tom Yum Chili Flavor","Style":"Pack","Country":"Thailand","Stars":"2","Top Ten":null},
  {"Review #":1,"Brand":"Westbrae","Variety":"Miso Ramen","Style":"Pack","Country":"USA","Stars":"0.5","Top Ten":null}
]"""
# should convert json string to dataFrame
# ระหว่างเรื่องของ index ฟังก์ชัน concat ควรใช้ ignore_index
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html
# output to sys.stdout

In [ ]:
import pandas as pd
import numpy as np 
# ,index_col='Review #
data_csv = pd.read_csv('https://raw.githubusercontent.com/ABChamp/python_programming_buu_class/master/data/hw-remen-ratings.csv',sep="|",names=['Review #','Brand','Variety','Style','Country','Stars','Top Ten'])
data_csv

Review #            Brand  ... Stars  Top Ten
0         2580        New Touch  ...  3.75      NaN
1         2579         Just Way  ...     1      NaN
2         2578           Nissin  ...  2.25      NaN
3         2577          Wei Lih  ...  2.75      NaN
4         2576   Ching's Secret  ...  3.75      NaN
...        ...              ...  ...   ...      ...
2565        15          Samyang  ...   2.5      NaN
2566        14          Samyang  ...   3.5      NaN
2567        13  Sapporo Ichiban  ...     5  2012 #4
2568        12  Sapporo Ichiban  ...   2.5      NaN
2569        11      Six Fortune  ...     2      NaN

[2570 rows x 7 columns]

In [ ]:

df = pd.read_json(jsonString)
# df.set_index('Review #',inplace=True)
df

Review #     Brand  ... Stars Top Ten
0        10     Smack  ...   1.5     NaN
1         9     Sutah  ...   2.0     NaN
2         8    Tung-I  ...   3.0     NaN
3         7   Ve Wong  ...   1.0     NaN
4         6     Vifon  ...   2.5     NaN
5         5     Vifon  ...   3.5     NaN
6         4   Wai Wai  ...   1.0     NaN
7         3   Wai Wai  ...   2.0     NaN
8         2   Wai Wai  ...   2.0     NaN
9         1  Westbrae  ...   0.5     NaN

[10 rows x 7 columns]

In [ ]:
quiz = data_csv.append(df, ignore_index=True)
data1 = quiz
quiz
# quiz1.to_csv (r'quiz1.csv', index = False, header=True)

Review #           Brand  ... Stars Top Ten
0         2580       New Touch  ...  3.75     NaN
1         2579        Just Way  ...     1     NaN
2         2578          Nissin  ...  2.25     NaN
3         2577         Wei Lih  ...  2.75     NaN
4         2576  Ching's Secret  ...  3.75     NaN
...        ...             ...  ...   ...     ...
2575         5           Vifon  ...   3.5     NaN
2576         4         Wai Wai  ...     1     NaN
2577         3         Wai Wai  ...     2     NaN
2578         2         Wai Wai  ...     2     NaN
2579         1        Westbrae  ...   0.5     NaN

[2580 rows x 7 columns]

In [ ]:
quiz.to_csv (r'quiz1.csv', index = False, header=True)

<b> ข้อที่ 2 </b> <br>

2.1 แสดงจำนวนแถวที่มีข้อมูลใน Top Ten <br>
<b> ผลลัพธ์เท่ากับ 37 แถว </b>

In [ ]:
dfgd = quiz['Top Ten'].dropna()
df1 = dfgd[dfgd!='\n'].count()
df1

37

In [ ]:
# ควรเช็คว่ามีข้อมูลอะไรบ้างที่ไม่เกี่ยวข้องกับ topten

2.2 เปลี่ยนชื่อประเทศในคอลัมภ์ Country จากชื่อเต็มเป็นชื่อย่อ

<img src="https://drive.google.com/uc?export=view&id=1JVKkLLaYfy3dGZXzzU9AO-yoDmPoQnff" width="500">

In [ ]:
country =  {
    'Japan': 'JP',
    'Taiwan': 'TW',
    'USA': 'USA',
    'India': 'IN',
    'South Korea': 'KR',
    'Singapore': 'SG',
    'Thailand': 'TH',
    'Hong Kong': 'HK',
    'Vietnam': 'VN',
    'Ghana': 'GH',
    'Malaysia': 'MY',
    'Indonesia': 'ID', 
    'China' : 'CN', 
    'Nigeria': 'NG', 
    'Germany': 'DE', 
    'Hungary': 'HU', 
    'Mexico': 'MX',
    'Fiji': 'FJ', 
    'Australia': 'AU', 
    'Pakistan': 'PK', 
    'Bangladesh': 'BD', 
    'Canada': 'CA', 
    'Nepal': 'NP',
    'Brazil': 'BR', 
    'UK': "UK", 
    'Myanmar': 'MM', 
    'Netherlands': 'NL', 
    'United States': 'US',
    'Cambodia': 'KH', 
    'Finland': 'FI', 
    'Sarawak': 'Sarawak', 
    'Philippines': 'PH', 
    'Sweden': 'SE',
    'Colombia': 'CO', 
    'Estonia': 'EE', 
    'Holland': 'Holland', 
    'Poland': 'PL', 
    'Dubai': 'Dubai'
}

In [ ]:
quiz3 = quiz
quiz3['Country'] = quiz3.Country.replace(country)
quiz3


Review #           Brand  ... Stars Top Ten
0         2580       New Touch  ...  3.75     NaN
1         2579        Just Way  ...     1     NaN
2         2578          Nissin  ...  2.25     NaN
3         2577         Wei Lih  ...  2.75     NaN
4         2576  Ching's Secret  ...  3.75     NaN
...        ...             ...  ...   ...     ...
2575         5           Vifon  ...   3.5     NaN
2576         4         Wai Wai  ...     1     NaN
2577         3         Wai Wai  ...     2     NaN
2578         2         Wai Wai  ...     2     NaN
2579         1        Westbrae  ...   0.5     NaN

[2580 rows x 7 columns]

2.3 สร้างคอลัมภ์ใหม่ชื่อว่า rounded_stars โดยที่นำคะแนนใน Stars เป็นชนิด float64

<img src="https://drive.google.com/uc?export=view&id=1L3hNQUx1psMLGVRkL81m4UPkZMPVgYj2" width="500">

In [ ]:
quiz['rounded_stars'] = quiz['Stars']

In [ ]:
# ข้อนีี้ควรเปลี่ยน unrated เป็น -1 (ที่จริงควร drop ออกเพราะมีผลแต่ตอนนี้ทำวิธีนี้ไปก่อน)

quiz10 = quiz
quiz10['rounded_stars'] = quiz10.rounded_stars.str.replace('Unrated', '-1').astype(float)
quiz10['rounded_stars'] = quiz10['rounded_stars'].astype(np.float64)
quiz10.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2577 entries, 0 to 2579
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Review #       2577 non-null   int64  
 1   Brand          2577 non-null   object 
 2   Variety        2577 non-null   object 
 3   Style          2575 non-null   object 
 4   Country        2577 non-null   object 
 5   Stars          2577 non-null   object 
 6   Top Ten        41 non-null     object 
 7   rounded_stars  2567 non-null   float64
dtypes: float64(1), int64(1), object(6)
memory usage: 181.2+ KB


In [ ]:
quiz.drop(quiz.loc[quiz['rounded_stars']=='Unrated'].index, inplace=True)
quiz['rounded_stars'] = quiz['rounded_stars'].astype(np.float64)
quiz

/usr/local/lib/python3.6/dist-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


Review #           Brand  ... Top Ten rounded_stars
0         2580       New Touch  ...     NaN          3.75
1         2579        Just Way  ...     NaN          1.00
2         2578          Nissin  ...     NaN          2.25
3         2577         Wei Lih  ...     NaN          2.75
4         2576  Ching's Secret  ...     NaN          3.75
...        ...             ...  ...     ...           ...
2575         5           Vifon  ...     NaN           NaN
2576         4         Wai Wai  ...     NaN           NaN
2577         3         Wai Wai  ...     NaN           NaN
2578         2         Wai Wai  ...     NaN           NaN
2579         1        Westbrae  ...     NaN           NaN

[2577 rows x 8 columns]

2.4 เปลี่ยนข้อความใน variety เป็นตัวเล็กทั้งหมด

In [ ]:

quiz['Variety'] = quiz['Variety'].str.lower()
quiz

Review #           Brand  ... Top Ten rounded_stars
0         2580       New Touch  ...     NaN          3.75
1         2579        Just Way  ...     NaN          1.00
2         2578          Nissin  ...     NaN          2.25
3         2577         Wei Lih  ...     NaN          2.75
4         2576  Ching's Secret  ...     NaN          3.75
...        ...             ...  ...     ...           ...
2575         5           Vifon  ...     NaN           NaN
2576         4         Wai Wai  ...     NaN           NaN
2577         3         Wai Wai  ...     NaN           NaN
2578         2         Wai Wai  ...     NaN           NaN
2579         1        Westbrae  ...     NaN           NaN

[2577 rows x 8 columns]

2.5 เปลี่ยนชื่อคอลัมภ์จาก Review # เป็น Review

In [ ]:
quiz.rename(columns={'Review #':'Review'})

Review           Brand  ... Top Ten rounded_stars
0       2580       New Touch  ...     NaN          3.75
1       2579        Just Way  ...     NaN          1.00
2       2578          Nissin  ...     NaN          2.25
3       2577         Wei Lih  ...     NaN          2.75
4       2576  Ching's Secret  ...     NaN          3.75
...      ...             ...  ...     ...           ...
2575       5           Vifon  ...     NaN          3.50
2576       4         Wai Wai  ...     NaN          1.00
2577       3         Wai Wai  ...     NaN          2.00
2578       2         Wai Wai  ...     NaN          2.00
2579       1        Westbrae  ...     NaN          0.50

[2577 rows x 8 columns]

<b> ข้อที่ 3 </b> <br>

3.1 หาคะแนนเฉลี่ย (mean) rounded_stars ในแต่ละ brand (ยี่ห้อ) <br>

<b> สุ่มผลลัพธ์จะได้ </b> <br>
Yum-Mie เท่ากับ 3.5 <br>
A-One  เท่ากับ 2.75 <br>
A-Sha Dry Noodle เท่ากับ 4.0673076923076925 <br>


In [ ]:
# 
quiz7 = quiz
# quiz7.drop(quiz7.loc[quiz7['rounded_stars']=='Unrated'].index, inplace=True)
# grp = quiz.groupby(['Brand'], axis=0)
quiz7.groupby(['Brand'])[['rounded_stars']].mean()
# grp['rounded_stars'].agg([np.sum, np.mean]) 

rounded_stars
Brand                          
1 To 3 Noodles         4.000000
7 Select               3.750000
7 Select/Nissin        3.500000
A-One                  2.750000
A-Sha Dry Noodle       4.067308
...                         ...
Yum Yum                3.750000
Yum-Mie                3.500000
Zow Zow                3.750000
iMee                   3.500000
iNoodle                3.000000

[355 rows x 1 columns]

3.2 นับจำนวนแถวจำนวน ramen ในแต่ละประเทศ แสดงผลออกมาเป็นแต่ละประเทศ

<b> สุ่มผลลัพธ์จะได้ </b> <br>
Japan = 352 <br>
Canada = 41 <br>
Thailand = 91 <br>

In [ ]:
quiz.groupby(['Country'])[['rounded_stars']].count()
# quiz.shape

rounded_stars
Country                     
Australia                 22
Bangladesh                 7
Brazil                     5
Cambodia                   5
Canada                    41
China                    169
Colombia                   6
Dubai                      3
Estonia                    2
Fiji                       4
Finland                    3
Germany                   27
Ghana                      2
Holland                    4
Hong Kong                137
Hungary                    9
India                     31
Indonesia                126
Japan                    352
Malaysia                 156
Mexico                    25
Myanmar                   14
Nepal                     14
Netherlands               15
Nigeria                    1
Pakistan                   9
Philippines               47
Poland                     4
Sarawak                    3
Singapore                109
South Korea              308
Sweden                     3
Taiwan                   223
Thailand                 188
UK                        69
USA                      321
United States              1
Vietnam                  105

3.3 นับจำนวนแถวที่มี style เป็นแบบ Pack ที่มี rounded_stars มากกว่าหรือเท่ากับค่าเฉลี่ย ( >= mean) <br>

<b> ผลลัพธ์ค่าเฉลี่ยจะได้เท่ากับ 3.6912475506205102
จำนวนแถวที่เป็น style เป็นแบบ Pack จะมีทั้งหมด 907 แถว </b>

In [ ]:
quiz33= quiz
sf = quiz33[quiz33['Style']=='Pack']
# sf.drop(sf.loc[sf['rounded_stars']=='Unrated'].index, inplace=True)
sda = sf[sf['rounded_stars'] >= float(3.6912475506205102)]
sda

Review #            Brand  ...  Top Ten rounded_stars
4         2576   Ching's Secret  ...      NaN          3.75
5         2575    Samyang Foods  ...      NaN          4.75
10        2570      Tao Kae Noi  ...      NaN          5.00
11        2569         Yamachan  ...      NaN          5.00
12        2568         Nongshim  ...      NaN          4.25
...        ...              ...  ...      ...           ...
2557        23           Nissin  ...      NaN          4.50
2558        22         Nongshim  ...      NaN          3.75
2559        21         Nongshim  ...      NaN          4.00
2561        19         Nongshim  ...      NaN          4.00
2567        13  Sapporo Ichiban  ...  2012 #4          5.00

[907 rows x 8 columns]

3.4 หาคะแนนเฉลี่ย (Mean) rounded_stars ของ ramen ที่มาจากประเทศไทย
<b> ผลลัพธ์จะเท่ากับ 3.3848167539267022</b>

In [ ]:
quiz

In [ ]:

quiz34 = quiz
fd = quiz34[(quiz34['Country']=='Thailand')][['rounded_stars']].mean()
fd

rounded_stars    3.384817
dtype: float64

<b> ข้อที่ 4 (Advance - ลองทำ)</b> <br>
4.1 หาคำว่า chicken ในฟิลล์ Variety ว่ามีทั้งหมดกี่ตัว
<b> ผลลัพธ์จะเท่ากับ 328 แถว </b>

<img src="https://drive.google.com/uc?export=view&id=1o5vMXO9fGcel_Xd-VEB3Hi3F5_NDarzF" width="500">



In [ ]:
gfg=quiz[quiz['Variety'].str.contains('chicken')]
gfg.shape

(328, 8)

4.2 แสดงแถวที่ได้รับรางวัล Top Ten โดยเรียงตามปีและลำดับ <br>
ผลลัพธ์ <br> 
<img src="https://drive.google.com/uc?export=view&id=18Mp_hQTXrtKEmOnzdirLcJDaeQjEKxiT" width="500">


In [ ]:
data = data1.sort_values(['Top Ten'],ascending=True).dropna()
data[data['Top Ten']!='\n']

Review #             Brand  ... Stars   Top Ten
2475       105           Indomie  ...     5   2012 #1
1972       608              Koka  ...     5  2012 #10
2533        47           Indomie  ...     5   2012 #2
2188       392            Nissin  ...     5   2012 #3
2567        13   Sapporo Ichiban  ...     5   2012 #4
2146       434         Mi Sedaap  ...   4.5   2012 #5
2189       391             Myojo  ...  4.25   2012 #6
2002       578          Nongshim  ...  4.75   2012 #7
2295       285              Doll  ...   4.5   2012 #9
1588       992       Prima Taste  ...     5   2013 #1
1109      1471              Mama  ...     5  2013 #10
1589       991       Prima Taste  ...     5   2013 #2
1865       715           Indomie  ...     5   2013 #3
1646       934          Nongshim  ...     5   2013 #4
2549        31             Myojo  ...     4   2013 #6
1757       823             Paldo  ...     5   2013 #9
1493      1087           MyKuali  ...     5   2014 #1
1278      1302              Mama  ...     5  2014 #10
1250      1330   Sapporo Ichiban  ...     5   2014 #4
1382      1198     Samyang Foods  ...     5   2014 #5
1397      1183             Paldo  ...     5   2014 #6
1230      1350             Mamee  ...     5   2014 #7
1309      1271       Prima Taste  ...     5   2014 #8
1272      1308          Nongshim  ...     5   2014 #9
1080      1500           MyKuali  ...     5   2015 #1
942       1638  A-Sha Dry Noodle  ...     5  2015 #10
995       1585            CarJEN  ...     5   2015 #4
1072      1508             Mamee  ...     5   2015 #6
963       1617           MyKuali  ...     5   2015 #7
1114      1466              Mama  ...     5   2015 #8
1059      1521          Maruchan  ...     5   2015 #9
633       1947       Prima Taste  ...     5   2016 #1
616       1964              MAMA  ...     5  2016 #10
673       1907       Prima Taste  ...     5   2016 #5
891       1689      Wugudaochang  ...     5   2016 #7
655       1925             Prima  ...     5   2016 #8
752       1828     Tseng Noodles  ...     5   2016 #9

[37 rows x 7 columns]